MmSupervisorPkg: Add plugin to scan binaries for access to x86 protected instructions

apop5 opened the pull request:
## Description
Post build plugin which scans mm supervisor binaries for accessing privileged instructions.

- [ ] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested
Ran on Q35 derivative using ClangPdb, ClangDwarf and VS2022

## Integration Instructions
Relies on BUILDREPORT_FILE being generated by build process to be able to determine StandaloneMm modules which need scanning.

Will have to deal with how to enable this when MM Supervisor is enabled. As of right now, when Mm Supv is present in the project, the build plugin will always run, even if its Mm Supervisor is not enabled.

apop5 committed:
Add plugin that scans mm supervisor binaries for msr accesses

apop5 commented:
example output:
Scanning: PiSmmCpuStandaloneMm
INFO - Read MSR  1A0h at 00000000000010E8
INFO - Write MSR 1A0h at 00000000000010E8
INFO - Read MSR  0C0000080h at 0000000000001101
INFO - Write MSR 0C0000080h at 0000000000001101
INFO - Read MSR  0C0000080h at 000000000000111E
INFO - Write MSR 0C0000080h at 000000000000111E
INFO - Read MSR  6A2h at 000000000000117A
INFO - Read MSR  6A0h at 0000000000001183
INFO - Read MSR  6A4h at 000000000000118C
INFO - Read MSR  6A8h at 0000000000001195
INFO - Write MSR 6A0h at 000000000000119E
INFO - Write MSR 6A2h at 00000000000011A9
INFO - Write MSR 6A4h at 00000000000011B7
INFO - Write MSR 6A8h at 00000000000011D8
INFO - Write MSR 6A8h at 000000000000127A
INFO - Write MSR 6A4h at 0000000000001283
INFO - Write MSR 6A0h at 000000000000128C
INFO - Write MSR 6A2h at 0000000000001295
INFO - Read MSR  1A0h at 00000000000012B7
INFO - Write MSR 1A0h at 00000000000012B7
INFO - Read MSR  0C0000080h at 00000000000016A8
INFO - Write MSR 0C0000080h at 00000000000016A8
INFO - Read MSR  1A0h at 0000000000002CE7
INFO - Write MSR 1A0h at 0000000000002CE7
INFO - Read MSR  0C0000080h at 0000000000002D00
INFO - Write MSR 0C0000080h at 0000000000002D00
INFO - Read MSR  0C0000080h at 0000000000002D1D
INFO - Write MSR 0C0000080h at 0000000000002D1D
INFO - Read MSR  6A2h at 0000000000002D6E
INFO - Read MSR  6A0h at 0000000000002D77
INFO - Read MSR  6A4h at 0000000000002D80
INFO - Read MSR  6A8h at 0000000000002D89
INFO - Write MSR 6A0h at 0000000000002DA1
INFO - Write MSR 6A2h at 0000000000002DAC
INFO - Write MSR 6A4h at 0000000000002DBA
INFO - Write MSR 6A8h at 0000000000002DDB
INFO - Write MSR 6A2h at 0000000000002E7E
INFO - Write MSR 6A8h at 0000000000002E94
INFO - Write MSR 6A4h at 0000000000002E9D
INFO - Write MSR 6A0h at 0000000000002EA6
INFO - Read MSR  1A0h at 0000000000002ED1
INFO - Write MSR 1A0h at 0000000000002ED1
INFO - Write MSR 6A2h at 0000000000002F17
INFO - Write MSR 141h at 0000000000002F25
INFO - Read MSR  0C0000080h at 0000000000002F9A
INFO - Read MSR  0C0000080h at 0000000000002FDA
INFO - Write MSR 0C0000080h at 0000000000002FDA
INFO - Read MSR  0C0000080h at 0000000000002FE6
INFO - Write MSR 0C0000080h at 0000000000002FE6
INFO - Read MSR  0C0000080h at 0000000000003079
INFO - Write MSR 0C0000080h at 0000000000003079

spbrogan commented on MmSupervisorPkg/Plugin/PostBuildScanMmStandalone_plug_in.json:
```diff
@@ -0,0 +1,12 @@
+## @file
```
you should probably change this to yaml which is allowed to have comments.  JSON isn't technically allowed to have comments so tools show errors.  

Also this looks like copy/paste from a different plugin.  Probably needs to be updated.

spbrogan commented on MmSupervisorPkg/Plugin/PostBuildScanMmStandalone_plug_in.json:
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+{
+    "scope": "global",
+    "name": "Post Build Scan Mm Standalone",
```
Better name needed :)

spbrogan commented on MmSupervisorPkg/Plugin/PostBuildScanMmStandalone.py:
```diff
@@ -0,0 +1,166 @@
+            List of disassembly lines.
+        """
+        try:
+            result = subprocess.run(
```
since you are using pytools please use RunCmd (https://github.com/tianocore/edk2-pytool-library/blob/23fb161210598927e315f60a03215397e5236e06/edk2toollib/utility_functions.py#L144)

apop5 replied:
fixed in latest

spbrogan commented on MmSupervisorPkg/Plugin/PostBuildScanMmStandalone_plug_in.json:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+{
+    "scope": "global",
```
should this be Windows only given it is using Windows tools

apop5 replied:
Changed to global-win, but it should probably be made its own scope to enable users to only enable this when mm_supv is used.

spbrogan commented on MmSupervisorPkg/Plugin/PostBuildScanMmStandalone.py:
```diff
@@ -0,0 +1,166 @@
+    instructions (rdmsr/wrmsr) in MM Standalone modules.
+    """
+
+    def do_post_build(self, thebuilder) -> int:
```
A little more structure here would be good.  Maybe a main function with various sub functions for each type of check.  I would expect over time we would be adding privileged instructions, port io, and other policy managed behavioral checks.

spbrogan reviewed (commented):
I like what is here, but a cleaner code layout would make it easier to expand over time.

apop5 committed:
Updating to focus on new goal: scanning mm supervisor binaries to ensure protected instructions are not used.

apop5 commented:
new example output:
`WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\MmSupervisorRing3Broker.efi
WARNING -       {'cli', 'out'}
WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\ReportStatusCodeRouterStandaloneMm.efi
WARNING -       {'out', 'in'}
WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\FirmwarePerformanceStandaloneMm.efi
WARNING -       {'in'}
WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\MmSupervisorErrorReport.efi
WARNING -       {'hlt', 'out'}
WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\CpuIo2StandaloneMm.efi
WARNING -       {'hlt', 'out', 'in'}
WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\FvbServicesStandaloneMm.efi
WARNING -       {'rdmsr', 'in'}
WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\FaultTolerantWriteStandaloneMm.efi
WARNING -       {'hlt', 'cli', 'sti', 'out', 'in'}`

apop5 committed:
Update to support Gcc/VS2022/Clang.

Change disassembly to include printed function names where instructions are encountered

apop5 commented:
Example output:
`WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_VS2022\X64\MmSupervisorPkg\Drivers\MmSupervisorRing3Broker\MmSupervisorRing3Broker\DEBUG\MmSupervisorRing3Broker.dll
WARNING -       {'EnableInterruptsAndSleep sti', 'EnableInterrupts sti', 'EnableInterruptsAndSleep hlt'}

WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_VS2022\X64\MsWheaPkg\MsWheaReport\Smm\MsWheaReportStandaloneMm\DEBUG\MsWheaReportStandaloneMm.dll
WARNING -       {'EnableInterruptsAndSleep sti', 'EnableInterrupts sti', 'EnableInterruptsAndSleep hlt'}

WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_VS2022\X64\MdeModulePkg\Universal\ReportStatusCodeRouter\Smm\ReportStatusCodeRouterStandaloneMm\DEBUG\ReportStatusCodeRouterStandaloneMm.dll
WARNING -       {'EnableInterruptsAndSleep sti', 'EnableInterrupts sti', 'EnableInterruptsAndSleep hlt'}

WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_VS2022\X64\MdeModulePkg\Universal\Acpi\FirmwarePerformanceDataTableSmm\FirmwarePerformanceStandaloneMm\DEBUG\FirmwarePerformanceStandaloneMm.dll
WARNING -       {'EnableInterruptsAndSleep sti', 'EnableInterrupts sti', 'EnableInterruptsAndSleep hlt'}`



WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\MmSupervisorPkg\Drivers\MmSupervisorRing3Broker\MmSupervisorRing3Broker\DEBUG\MmSupervisorRing3Broker.dll
WARNING -       {'_ModuleEntryPointWorker hlt'}

WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\UefiCpuPkg\CpuIo2Smm\CpuIo2StandaloneMm\DEBUG\CpuIo2StandaloneMm.dll
WARNING -       {'_ModuleEntryPointWorker hlt'}

WARNING - Protected Instructions found in C:/Code/qemu\Build\QemuQ35Pkg\DEBUG_CLANGDWARF\X64\QemuQ35Pkg\QemuFlashFvbServicesRuntimeDxe\FvbServicesStandaloneMm\DEBUG\FvbServicesStandaloneMm.dll
WARNING -       {'InternalMemEncryptSevStatus rdmsr', '_ModuleEntryPointWorker hlt'}

kuqin12 commented:
> We can, but I am still not happy with the llvm-dumpbin and dumpbin from llvm/gcc. They are bad about attempting to decode jumptables that the code generates, and throwing false positives depending on what was in the tables.

Hmm, can we get this in so that we can start to use it for MSVC centric platforms? This is a useful tool we probably should spread it out for greater adoption. I agree with your concerns on the llvm, but we can gradually improve it on the go?

apop5 committed:
Updated with Readme

apop5 committed:
Markdown lint and spelling

apop5 committed:
add spell check exceptions for valid assembly instructions, and linux command line options

apop5 committed:
Merge branch 'main' into personal/apop5/scanningbinaries
